File: python/src/agent_framework/enterprise/quota_in_memory.py
"""In-memory quota store."""

from agent_framework.enterprise.quota import QuotaStore, QuotaCheck, QuotaResult


class InMemoryQuotaConfig:
    def __init__(self, metric: str, limit: float, tenant_id: str | None = None):
        self.tenant_id = tenant_id or "default"
        self.metric = metric
        self.limit = limit


class InMemoryQuotaStore(QuotaStore):
    def __init__(self, configs: list[InMemoryQuotaConfig] | None = None):
        self._limits: dict[str, float] = {}
        self._usage: dict[str, float] = {}
        for c in configs or []:
            self.set_limit(c.tenant_id, c.metric, c.limit)

    def set_limit(self, tenant_id: str, metric: str, limit: float) -> None:
        self._limits[f"{tenant_id}:{metric}"] = limit

    def _key(self, check: QuotaCheck) -> str:
        return f"{check.tenant_id or 'default'}:{check.metric}"

    async def check(self, request: QuotaCheck) -> QuotaResult:
        k = self._key(request)
        limit = self._limits.get(k)
        if limit is None:
            return QuotaResult(allowed=True)
        used = self._usage.get(k, 0)
        remaining = max(0, limit - used)
        return QuotaResult(
            allowed=used + request.amount <= limit,
            remaining=remaining,
            limit=limit,
        )

    async def consume(self, request: QuotaCheck) -> QuotaResult:
        k = self._key(request)
        limit = self._limits.get(k)
        if limit is None:
            return QuotaResult(allowed=True)
        self._usage[k] = self._usage.get(k, 0) + request.amount
        used = self._usage[k]
        remaining = max(0, limit - used)
        return QuotaResult(allowed=used <= limit, remaining=remaining, limit=limit)

File: packages/enterprise/src/quota-in-memory.ts
import type { QuotaStore, QuotaCheck, QuotaResult } from "./quota.js";

export interface InMemoryQuotaConfig {
  tenantId?: string;
  metric: string;
  limit: number;
  windowMs?: number;
}

/**
 * In-memory quota store; production should use Redis or DB for distributed consistency.
 */
export class InMemoryQuotaStore implements QuotaStore {
  private limits = new Map<string, number>();
  private usage = new Map<string, number>();

  constructor(configs: InMemoryQuotaConfig[] = []) {
    for (const c of configs) {
      this.setLimit(c.tenantId ?? "default", c.metric, c.limit);
    }
  }

  setLimit(tenantId: string, metric: string, limit: number): void {
    this.limits.set(`${tenantId}:${metric}`, limit);
  }

  private key(check: QuotaCheck): string {
    return `${check.tenantId ?? "default"}:${check.metric}`;
  }

  async check(request: QuotaCheck): Promise<QuotaResult> {
    const k = this.key(request);
    const limit = this.limits.get(k);
    if (limit === undefined) return { allowed: true };
    const used = this.usage.get(k) ?? 0;
    const remaining = Math.max(0, limit - used);
    return {
      allowed: used + request.amount <= limit,
      remaining,
      limit,
    };
  }

  async consume(request: QuotaCheck): Promise<QuotaResult> {
    const k = this.key(request);
    const limit = this.limits.get(k);
    if (limit === undefined) return { allowed: true };
    const used = (this.usage.get(k) ?? 0) + request.amount;
    this.usage.set(k, used);
    const remaining = Math.max(0, limit - used);
    return {
      allowed: used <= limit,
      remaining,
      limit,
    };
  }
}
